### PR TITLE
Feature/multiindex dataframes

### PR DESCRIPTION
**Pull request by ryan-kipawa**
MIKE IO 1D currently reads to pandas dataframes that contain a single column name. The column name is a string representation of a QueryData object, which concatenates several attributes with a delimiter. There are two current limitations of this approach which this PR aims to solve:

1. The column names are not a 1-1 mapping to MIKE 1D results, and duplicates have been observed.
2. The column names are cumbersome to handle, especially when trying to do operations along a specific axis (e.g. chainage).

This PR fundamentally introduces two new concepts: a new class called TimeseriesId, and pandas dataframes with hierarchical indices. TimeseriesId is a dataclass that uniquely identifies timeseries results. It is agnostic, allowing to/from conversion from several existing objects (e.g. QueryData and ResultQuantity). Additionally, it can be converted to/from a pandas hierarchical index, where each level matches a dataclass field.

A multiindex dataframe allows usecases like:
```python
res = Res1D('my_river.res1d')
df = res.read()

# Plot the variation of 'Discharge' along the chainage of 'my_river'.
(
    df
     .T.query("name=='my_river' and quantity=='Discharge'").T
     .T.groupby("chainage").max().T    # grab max discharge if there's two at one chainage
     .describe().T
     .plot(y=["min", "25%", "mean", "max"])
)
``` 

Hierarchical indices is an advanced pandas feature. Options should be provided to allow reading to a non-hierarchical dataframe, as well as helper methods for doing common query/groupby operations.

### CODE REVIEW

**Comment by ecomodeller** [comment]
> (
>     df
>      .T.query("name=='my_river' and quantity=='Discharge'").T
>      .T.groupby("chainage").max().T    # grab max discharge if there's two at one chainage
>      .describe().T
>      .plot(y=["min", "25%", "mean", "max"])
> )

What's up with all the `.T`'s?

File: mikeio1d/res1d.py [hunk 1]
```diff
@@ -130,15 +137,72 @@ def __repr__(self):
...
+        List of TimeSeriesId objects.
+        """
+        if queries is None:
+            return self.result_network.queue
+
+        timeseries_ids = self._validate_queries_as_timeseries_ids(queries)
+
+        return timeseries_ids
+
+    def _validate_queries_as_timeseries_ids(
```
Comment (gedaskir) [inline mikeio1d/res1d.py:164]: I suggest to move the contents of the method to result_query module. I think the idea with the Res1D class, that it is collecting high level method call and has as little as possible particularities.

File: mikeio1d/res1d.py [hunk 2]
```diff
@@ -302,7 +390,13 @@ def save(self, file_path):
         self.data.Connection.FilePath.Path = file_path
         self.data.Save()
 
-    def extract(self, file_path, queries=None, time_step_skipping_number=1, ext=None):
```
Comment (gedaskir) [inline mikeio1d/res1d.py:305]: While we are at it add type hints for other arguments?

File: mikeio1d/result_network/result_global_data.py [hunk 3]
```diff
@@ -36,8 +51,7 @@ def set_quantities(self):
 
     def add_to_result_quantity_maps(self, quantity_id, result_quantity):
         """Add global data result quantity to result quantity maps."""
-        query = QueryDataGlobal(quantity_id)
```
Comment (gedaskir) [inline mikeio1d/result_network/result_global_data.py:39]: Though it is only a removal of a single line, but for me it shows that introduction of TimeSeriesId and related refactoring is a very good thing :-).

File: mikeio1d/result_network/result_quantity.py [hunk 4]
```diff
@@ -73,12 +96,21 @@ def to_txt(self, file_path, time_step_skipping_number=1):
...
 
     def get_data_entry_net(self):
         """Get DataEntryNet corresponding to ResultQuantity."""
         return DataEntryNet(self.data_item, self.element_index)
+
+    @property
+    def timeseries_id(self) -> TimeSeriesId:
+        """TimeSeriesId corresponding to ResultQuantity."""
+        if self._timeseries_id is None:
+            raise ValueError(
```
Comment (gedaskir) [inline mikeio1d/result_network/result_quantity.py:113]: Use `message = "ResultQuantity must be added to a ResultNetwork before TimeSeriesId can be accessed."` and then `raise ValueError(message)`?

File: mikeio1d/result_query/query_data.py [hunk 5]
```diff
@@ -1,8 +1,16 @@
...
+
 import numpy as np
 
 from ..custom_exceptions import NoDataForQuery
 from ..custom_exceptions import InvalidQuantity
 from ..various import NAME_DELIMITER
+from ..quantities import TimeSeriesId
 
 
 class QueryData:
```
Comment (gedaskir) [inline mikeio1d/result_query/query_data.py:16]: I guess we can also abstract (ABC) this one as well?

File: mikeio1d/result_query/query_data.py [hunk 6]
```diff
@@ -35,20 +43,14 @@ def _validate(self):
         if self.name is not None and not isinstance(self.name, str):
             raise TypeError("Argument 'name' must be either None or a string.")
 
-    def add_to_data_entries(self, res1d, data_entries, column_names):
```
Comment (gedaskir) [inline mikeio1d/result_query/query_data.py:38]: Why we can get rid of this one?

File: mikeio1d/result_reader_writer/result_reader.py [hunk 7]
```diff
@@ -45,6 +56,16 @@ class ResultReader:
         Flag specifying to add chainage into data frame column label.
     """
 
+    class ColumnMode(str, Enum):
```
Comment (gedaskir) [inline mikeio1d/result_reader_writer/result_reader.py:59]: I suggest to take out the enum one level up. I think the nested classes/enums like that are fine if they are really private helper classes.

File: mikeio1d/result_reader_writer/result_reader.py [hunk 8]
```diff
@@ -45,6 +56,16 @@ class ResultReader:
...
+        """Specifies the type of column index of returned DataFrames."""
+
+        ALL = "all"
+        """Uses a column MultiIndex with all possible metadata."""
+        QUERY = "query"
```
Comment (gedaskir) [inline mikeio1d/result_reader_writer/result_reader.py:64]: Suggestion: we could also call it STRING or LABEL.

File: mikeio1d/result_reader_writer/result_reader.py [hunk 9]
```diff
@@ -81,6 +102,14 @@ def __init__(
 
         self.quantities = [quantity.Id for quantity in self.data.Quantities]
 
+        self.column_mode: ResultReader.ColumnMode = self.ColumnMode.ALL
```
Comment (gedaskir) [inline mikeio1d/result_reader_writer/result_reader.py:105]: Lets use ColumnMode.QUERY as a default. Now the res1d.ipynb also gets broke with ColumnMode.ALL as a default.

File: mikeio1d/result_reader_writer/result_reader.py [hunk 10]
```diff
@@ -142,10 +171,41 @@ def _add_catchment(self, catchment_id):
 
     # endregion File loading
 
-    def read(self, queries=None):
+    def read(
```
Comment (gedaskir) [inline mikeio1d/result_reader_writer/result_reader.py:174]: Lets make ResultReader also abstract?

File: mikeio1d/result_reader_writer/result_reader_copier.py [hunk 11]
```diff
@@ -70,26 +95,55 @@ def create_data_frame(self, data_entries, column_names):
...
             if not self.is_data_set_included(data_set):
                 continue
 
             for data_item in data_set.DataItems:
+                data_item = impl(data_item)
                 for i in range(data_item.NumberOfElements):
                     data_entry = DataEntryNet(data_item, i)
                     data_entries.Add(data_entry)
 
-                    column_name = self.get_column_name(data_set, data_item, i)
```
Comment (gedaskir) [inline mikeio1d/result_reader_writer/result_reader_copier.py:92]: Suggestion: keep `self.get_column_name` and rename to `self.get_timeseries_id` and put the contents with the while loop in it. Make _timeseires_ids and _timeseries_ids_set "private" instance variables.

File: mikeio1d/result_reader_writer/result_reader_copier.py [hunk 12]
```diff
@@ -70,26 +95,55 @@ def create_data_frame(self, data_entries, column_names):
...
+        columns = self.create_column_index(timeseries_ids, column_mode=column_mode)
+
+        df = pd.DataFrame(data_array, index=self.time_index, columns=columns)
 
         self.update_time_quantities(df)
 
         return df
 
-    def get_all_data_entries_and_column_names(self):
+    def create_column_index(
```
Comment (gedaskir) [inline mikeio1d/result_reader_writer/result_reader_copier.py:106]: Is this also relevant for result_reader_query and should go to the parent class?

File: mikeio1d/result_reader_writer/result_reader_creator.py [hunk 13]
```diff
@@ -22,8 +30,8 @@ def create(
         catchments=None,
         col_name_delimiter=NAME_DELIMITER,
         put_chainage_in_col_name=True,
-    ):
-        reasult_readers = {
+    ) -> ResultReader:
+        reasult_readers: Dict[ResultReaderType, ResultReader] = {
```
Comment (gedaskir) [inline mikeio1d/result_reader_writer/result_reader_creator.py:34]: I think here type hint is too much :-).

File: mikeio1d/result_reader_writer/result_writer.py [hunk 14]
```diff
@@ -90,7 +91,14 @@ def set_values_indexed(self, time_index, values, data_item, element_index):
         if isinstance(values, pd.DataFrame):
             values = values.iloc[:, 0]
 
-        for time in time_index:
-            value = float(values[time])
-            i = res1d_time_index.get_loc(time)
-            data_item.TimeData.SetValue(i, element_index, value)
+        length = len(values)
+        if length != len(time_index):
```
Comment (gedaskir) [inline mikeio1d/result_reader_writer/result_writer.py:95]: Introduce _validate_time_index method.

File: tests/test_res1d_catchments.py [hunk 15]
```diff
@@ -167,10 +168,10 @@ def test_catchment_attributes(test_file):
     catchments.c_22_8_8.TotalRunOff.add()
     df = res1d.read()
 
-    actual_max = round(df["TotalRunOff:20_2_2"].max(), 3)
```
Comment (gedaskir) [inline tests/test_res1d_catchments.py:170]: I think the tests with query label should still survive.

**Review by gedaskir — COMMENTED** [review]
This is my initial review of TimeSeriesId functionality. It is a very good improvement, especially the multiindices. Also lets add `ColumnMode.COMPACT`, which removes all irrelevant multiindex entries. For example, for `Catchment` kind data the chainage, tag, etc. will be irrelevant.

File: mikeio1d/quantities/__init__.py [hunk 16]
```diff
@@ -0,0 +1,4 @@
+from .timeseries_id import TimeSeriesId  # noqa: F401
+from .timeseries_id import TimeSeriesIdGroup  # noqa: F401
+
+all = ["TimeSeriesId", "TimeSeriesIdGroup"]
```
[the remark this answers is not in the document]
Comment (ecomodeller) [reply mikeio1d/quantities/__init__.py:4]: I don't think we *need* `__all__`.

But this convention is used in most packages for reasons found here: https://docs.python.org/3/tutorial/modules.html#importing-from-a-package

**Review by gedaskir — APPROVED** [review]

**Pull request merged**